pulsar-broker/src/main/java/org/apache/pulsar/broker/transaction/buffer/impl/SnapshotSegmentAbortedTxnProcessorImpl.java

```diff
@@ -0,0 +1,599 @@
+            });
+        }
+
+        private CompletableFuture<Void> deleteSnapshotSegment(PositionImpl positionNeedToDelete) {
+            long sequenceIdNeedToDelete = indexes.get(positionNeedToDelete).getSequenceID();
+            return snapshotSegmentsWriterFuture
+                    .thenCompose(writer -> writer.deleteAsync(buildKey(sequenceIdNeedToDelete), null))
+                    .thenRun(() -> {
+                        if (log.isDebugEnabled()) {
+                            log.debug("[{}] Successes to delete the snapshot segment, "
+                                            + "whose sequenceId is [{}] and maxReadPosition is [{}]",
+                                    this.topic.getName(), this.sequenceID, positionNeedToDelete);
+                        }
+                        abortTxnSegments.remove(positionNeedToDelete);
+                        //The process will check whether the snapshot segment is null, and update index when recovered.
+                        indexes.remove(positionNeedToDelete);
+                        //Keep index snapshot and update index
+                        updateSnapshotIndex(persistentSnapshotIndexes.getSnapshot(),
+                                indexes.values().stream().toList());
+                    }).exceptionally(e -> {
+                        log.warn("[{}] Failed to delete the snapshot segment, "
+                                        + "whose sequenceId is [{}] and maxReadPosition is [{}]",
+                                this.topic.getName(), this.sequenceID, positionNeedToDelete, e);
+                        return null;
+                    });
+        }
+
+        //Update the indexes with the giving index snapshot and index list in the transactionBufferSnapshotIndexes.
+        private CompletableFuture<Void> updateSnapshotIndex(TransactionBufferSnapshotIndexesMetadata snapshotSegment,
+                                                            List<TransactionBufferSnapshotIndex> indexList) {
+            TransactionBufferSnapshotIndexes snapshotIndexes = new TransactionBufferSnapshotIndexes();
+            return snapshotIndexWriterFuture
+                    .thenCompose((indexesWriter) -> {
+                        snapshotIndexes.setIndexList(indexList);
+                        snapshotIndexes.setSnapshot(snapshotSegment);
+                        return indexesWriter.writeAsync(topic.getName(), snapshotIndexes);
+                    })
+                    .thenRun(() -> {
+                        persistentSnapshotIndexes = snapshotIndexes;
+                        lastSnapshotTimestamps = System.currentTimeMillis();
+                    })
+                    .exceptionally(e -> {
+                        log.error("[{}] Failed to update snapshot segment index", snapshotIndexes.getTopicName(), e);
+                        return null;
+                    });
+        }
+
+        private CompletableFuture<Void> clearSnapshotSegmentAndIndexes() {
```

Comment:
Create a delete task for each segment, and update the index within each delete task. This method costs too much.

Reply:
>> Create a delete task for each segment, and update the index within each delete task. This method costs too much.

> In factly, this logic in clearSnapshotSegmentAndIndexes is executed in a single task.
There does not open new tasks for each segment deleting and index updating.

Can we just batch-delete The Segments and only update The Indexes once?

Reply:
Do you mean to add the `batch delete ` method to the `systemTopic::Writer`?